lib/internal/modules/cjs/loader.js

```diff
@@ -240,7 +241,21 @@ Module._findPath = function(request, paths, isMain) {
     var rc = stat(basePath);
     if (!trailingSlash) {
       if (rc === 0) {  // File.
-        if (preserveSymlinks && !isMain) {
+        if (!isMain) {
+          if (preserveSymlinks) {
+            filename = path.resolve(basePath);
+          } else {
+            filename = toRealPath(basePath);
+          }
+        } else if (preserveSymlinksMain) {
+          // For the main module, we use the preserveSymlinksMain flag instead
+          // mainly for backward compatibility, as the preserveSymlinks flag
+          // historically has not applied to the main module.  Most likely this
+          // was intended to keep .bin/ binaries working, as following those
+          // symlinks is usually required for the imports in the corresponding
+          // files to resolve; that said, in some use cases following symlinks
+          // causes bigger problems which is why the preserveSymlinksMain option
+          // is needed.
```

Comment:
I noticed a bit lower in `lib/internal/modules/cjs/loader.js` that `tryFile` has a similar `preserveSymlinks && !isMain` check. Should it also add the `preserveSymlinksMain` juggle there?

https://github.com/nodejs/node/blob/452d73df23f3283797f1e1099a1fddc635643d45/lib/internal/modules/cjs/loader.js#L193-L199

Reply:
hm, maybe?  I'm really not familiar with this code.  --preserve-symlinks-main did pass the test cases I wrote for it, though maybe those aren't exhaustive enough?  I wonder if there's something unique about the way the main file is loaded that it doesn't hit the tryFile code path?  Just a guess.

If we can find a case where the code misbehaves as written then it's definitely something that we should fix.

Reply:
reading a little closer... I *think* that node is supposed to be able to figure out that if `asdf.js` exists, then `node asdf` supposed to mean `node asdf.js`.  It looks like what you've noticed is that the code for that doesn't handle --preserve-symlinks-main like it would handle --preserve-symlinks, which does seem like a bug, though at least an easy to work around one (just remember to specify the extension on the command line)